### PR TITLE
sk-es-alert-user

### LINKED ISSUE #8: 6. As a user, I want to be alerted when I’m entering an item that is the same as something already on my list so I can avoid duplicate items
AC:

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

### PR DESCRIPTION
**Pull request by SaraSweetie**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 
## To Do
- [x] work with group on how to store data (collections and docs) in firebase
- [x] Clean out test data
- [x] get name field(s) from database
- [x] get the name from the form input field
- [x] compare name (shopping list item) from database with name user entered
- [x] normalize the 2 strings that are being compared
- [x] Use .toLower() - make all lowercase, .trim() - remove beginning and end spaces, and .replace('NEED regular expression', '') - replace punctuation and special characters with nothing
- [x] if input name doesn't match save a new item to the shopping list
- [x] else show the user an error message

## Description

This PR adds new functionality to the `AddItemForm` component. The `AddItemForm` now has the ability to display an error message when a user inputs an item that has already been saved to the database. 

### Steps to accomplish this feature:

We've updated the state of `AddItemForm` to hold a `name_normalize` property. This was done to differentiate between the user input: `name` from `name_normalize` when saved to the database. The name_normalize is a feature that accounts for any whitespace, capitalization, and special characters using a regex pattern. The regex pattern is used on the incoming item added by the user.

Once an item is added, it's been submitted to the database where it can be retrieved to be filtered. We are able to do this by the use of the `FirestoreCollection` tag within our form. Firestore collection is able to access the `items` collection and create a simple query to filter the normalized name to the name saved to the database. The filtering will generate an error message with: "`item name` has already been added to the shopping list" if the input is already in the database. 

## Related Issue

closes #8 

## Acceptance Criteria

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before
Screen Shot 2020-04-23 at 6 22 10 PM

### After
Screen Shot 2020-04-23 at 6 17 24 PM

## Testing Steps / QA Criteria

1. Git pull the branch using `git pull origin sk-es-alert-user` and `git checkout sk-es-alert-user` to switch to the branch.
2. Run npm install to add dependencies, then npm start to run the app.
3. Type an item into the name input field
4. Select a scheduled time and hit submit.
5. To test If It's working properly: Select the same item again, you can even add a special character or capital letter. 
6. An error message should appear. 

**Note:** `apple`, `Apple1`, and `Peanut Butter` are currently saved in the database and will present an error message if typed into the name field.

### CODE REVIEW

File: src/components/FormAlert.js
Comment (stevelikesmusic) [inline src/components/FormAlert.js]: You can always use a short circuit pattern for a `render this or nothing`:

```jsx
return someCondition && <p>Hey, the condition was true!</p>;
```

File: src/components/AddItemForm.js
Comment (stevelikesmusic) [inline src/components/AddItemForm.js]: You may not even need a separate component for this unless we were going to use a single component for any errors/messages across the app.

File: src/components/AddItemForm.js
Comment (skillitzimberg) [inline src/components/AddItemForm.js]: Do you mean you want to do something like `{this.state.name}` in place of "This item"? Or is this a stretch goal?

Comment (SaraSweetie) [reply src/components/AddItemForm.js]: Yes that's what we wanted to do. We moved this component up and deleted this file. So now we have access to this.state.name

File: src/components/AddItemForm.js
Comment (skillitzimberg) [inline src/components/AddItemForm.js]: This is not a blocker in any way, and there are probably different schools of thought about this, but I think the `normalizeUserInput` method definition should be moved up to just after the `handleChange` method definition. That way a reader doesn't have to look around for `normalizeUserInput` when they see that `handleChange` calls it.

File: src/components/FormAlert.js
Comment (skillitzimberg) [inline src/components/FormAlert.js]: Reminding myself as much as anyone else here . . . remove `console.log`'s and commented code. Why is this so hard to remember in the moment?

File: src/components/AddItemForm.js
Comment (skillitzimberg) [inline src/components/AddItemForm.js]: Not a blocker. I wonder if this should be `null` or an empty `p` tag . . . ? Just because it's kind of confusing that `message` is initialized as an empty `div` then is changed to a `p` when an alert is made. (My bad.)

File: src/components/AddItemForm.js
Comment (skillitzimberg) [inline src/components/AddItemForm.js]: Not a blocker; just something I realized while reading through this component.

Because the normalized input is being set in state as the user types, `this.state.name_normalized` is available and this method doesn't need to be called again.

`this.state.name_normalized` can be used instead on line 89: filter={['name_normalized', '==', this.state.name_normalized]}

File: src/components/AddItemForm.js
Comment (skillitzimberg) [inline src/components/AddItemForm.js]: This is a smart idea! Slick!

File: src/components/FormAlert.js
Comment (skillitzimberg) [reply src/components/FormAlert.js]: I just now figured out how to use @stevelikesmusic's suggestion in this context:

```
          render={({ data }) => {
            return data.length > 0 ? (
              <p>
                {data[0].name} has already been added to your shopping list.
              </p>
            ) : null;
          }}
```

**Review by skillitzimberg — APPROVED** [review]
Approved! Looks great!

File: src/components/FormAlert.js
Comment (SaraSweetie) [inline src/components/FormAlert.js]: this file is no longer being used and has been deleted

File: src/components/FormAlert.js
Comment (SaraSweetie) [reply src/components/FormAlert.js]: This whole `FormAlert.js` is not being used lol. I deleted the file. This was moved up a level into `AddItemForm.js`

File: src/components/AddItemForm.js
Comment (SaraSweetie) [reply src/components/AddItemForm.js]: I like this thought. Anything to help the readability of the code is good!

File: src/components/AddItemForm.js
Comment (SaraSweetie) [reply src/components/AddItemForm.js]: this was fixed with the suggestion to `short circuit pattern for a render this or nothing`. Thanks, Scott I understand what that meant now!

File: src/components/AddItemForm.js
Comment (SaraSweetie) [reply src/components/AddItemForm.js]: Wouldn't that be comparing the same thing though?? `this.state.name_normalized` is what is saved in state. The first parameter from the filter is also pulling the `name_normalized` from state. `normalizedInput` aka `this.normalizeUserInput(this.state.name)` is the user's input being run through the function as they are typing.

I did remove the variable and just use the function in the filer `{['name_normalized', '==', this.normalizeUserInput(this.state.name)]}`

File: src/components/AddItemForm.js
Comment (stevelikesmusic) [inline src/components/AddItemForm.js]: Because this is saved to state, we are calling `firebase` with each `onChange` event on the input. Depending on how much is typed, this can easily take us over our limit in `firebase`. If we stick with approach of using `FirestoreCollection` here, how might you limit the number of requests being made? Would it be fine to only send a request before actually sending the add request? 

Also, this query is running for _all_ lists. If I create a new list and try to add Peanut Butter, [I get the alert](https://cdn.zappy.app/03a6551a275756564dde1c76a1e9a415.mp4)

File: src/components/AddItemForm.js
Comment (stevelikesmusic) [inline src/components/AddItemForm.js]: Probably don't need to keep this in state. We can always use `this.props.userToken` in the `onSubmit`

File: src/lib/testList.js
Comment (stevelikesmusic) [inline src/lib/testList.js:17]: Is this used for anything?

File: src/components/AddItemForm.js
Comment (stevelikesmusic) [inline src/components/AddItemForm.js]: Not a blocker, you'll often use a short-hand for object keys that map to variables of the same name: 
```js
const normalized_name = this.normalize();
this.setState({
  name: value,
  name_normalized,
});
```

or I might just leave out the variable:
```js
this.setState({
  name: value,
  name_normalized: this.normalize(),
});

```

**Review by stevelikesmusic — CHANGES_REQUESTED** [review]
This is a clever approach! There are a few concerns, but I think this should work with a few tweaks. 

The over fetching is one thing to try to clean up. We want to avoid sending too many unnecessary requests to firebase. I'm going to block the merge for that, because sending too many requests to firebase could lock our account for a time.

How can we adjust the filter on the firebase query so it only looks for the entered item in the users current list?

It looks like the root view isn't working. I don't see any of the items I've added in my shopping list.

Another question: should we still be able to save the item if we know it's a duplicate?

File: src/App.js
Comment (stevelikesmusic) [inline src/App.js]: I forgot to delete `withFirestore`. Could you remove this and its import before merging?

**Review by stevelikesmusic — APPROVED** [review]
This all looks good. Thanks for incorporating the changes @espain16! 👏

**Pull request merged**
